pkg/keymanagementprovider/azurekeyvault/provider.go

```diff
@@ -233,33 +278,66 @@ func (s *akvKMProvider) GetKeys(ctx context.Context) (map[keymanagementprovider.
 
 		// fetch the key object from Key Vault
 		startTime := time.Now()
-		keyResponse, err := s.keyKVClient.GetKey(ctx, keyVaultKey.Name, keyVaultKey.Version)
-		if err != nil {
-			return nil, nil, fmt.Errorf("failed to get key objectName:%s, objectVersion:%s, error: %w", keyVaultKey.Name, keyVaultKey.Version, err)
+		// if versionHistoryLimit is not set, set it to default value 1
+		if keyVaultKey.VersionHistoryLimit == 0 {
+			keyVaultKey.VersionHistoryLimit = versionHistoryLimitDefault
 		}
-		keyBundle := keyResponse.KeyBundle
-		isEnabled := *keyBundle.Attributes.Enabled
-		// if version is set as "" in the config, use the version from the key bundle
-		keyVaultKey.Version = getObjectVersion(string(*keyBundle.Key.KID))
-
-		if !isEnabled {
-			startTime := time.Now()
-			lastRefreshed := startTime.Format(time.RFC3339)
-			properties := getStatusProperty(keyVaultKey.Name, keyVaultKey.Version, lastRefreshed, isEnabled)
-			keysStatus = append(keysStatus, properties)
-			mapKey := keymanagementprovider.KMPMapKey{Name: keyVaultKey.Name, Version: keyVaultKey.Version, Enabled: isEnabled}
-			keymanagementprovider.DeleteKeyFromMap(s.resource, mapKey)
-			continue
+
+		var versionHistory []struct {
+			Version string
+			Created time.Time
 		}
 
-		publicKey, err := getKeyFromKeyBundle(keyBundle)
-		if err != nil {
-			return nil, nil, fmt.Errorf("failed to get key from key bundle:%w", err)
+		keyVersionPager := s.keyKVClient.NewListKeyVersionsPager(keyVaultKey.Name, nil)
+		for keyVersionPager.More() {
+			pager, err := keyVersionPager.NextPage(ctx)
+			if err != nil {
+				return nil, nil, fmt.Errorf("failed to get key versions for objectName:%s, error: %w", keyVaultKey.Name, err)
+			}
+			for _, key := range pager.Value {
+				versionHistory = append(versionHistory, struct {
+					Version string
+					Created time.Time
+				}{Version: key.KID.Version(), Created: *key.Attributes.Created})
+			}
+		}
+
+		// sort the version history by created time
+		sortVersionHistory(versionHistory)
+		sortedVersionHistory := GetSortedVersions(versionHistory)
+
+		// if versionHistoryLimit is greater than the number of versions, set it to the number of versions
+		if keyVaultKey.VersionHistoryLimit > len(sortedVersionHistory) {
+			keyVaultKey.VersionHistoryLimit = len(sortedVersionHistory)
+		}
+
+		// get the latest version of the certificate up to the limit
+		for _, version := range sortedVersionHistory[len(sortedVersionHistory)-keyVaultKey.VersionHistoryLimit:] {
+			keyResponse, err := s.keyKVClient.GetKey(ctx, keyVaultKey.Name, version)
+			if err != nil {
+				return nil, nil, fmt.Errorf("failed to get key objectName:%s, objectVersion:%s, error: %w", keyVaultKey.Name, version, err)
+			}
+			keyBundle := keyResponse.KeyBundle
+			isEnabled := *keyBundle.Attributes.Enabled
```

Comment:
nit: is `keyBundle` a pointer? if so we might consider a nil check

Reply:
`keyBundle` isn't a pointer but the Attributes is. I can add a nil check for that field. Should I also add a nil check for the `SecretBundle` field in the GetCertificates method?

Reply:
Added changes for the `SecretBundle` as well. :) fc8a17ceb14f6358e43225b8627805aec89421b3